pkgpanda/__init__.py

```diff
@@ -77,6 +88,65 @@ def stop_all(self):
                 if ex.returncode != 5:
                     raise
 
+    def remove_staged_unit_files(self):
+        """Remove staged unit files created by Systemd.stage_new_units()."""
+        for filename in os.listdir(self.__base_systemd):
+            if filename.endswith(self.new_unit_suffix):
+                os.remove(os.path.join(self.__base_systemd, filename))
+
+    def stage_new_units(self, new_wants_dir):
+        """Prepare new systemd units for activation.
+
+        Unit files targeted by the symlinks in new_wants_dir are copied to a temporary location in the base systemd
+        directory, and the symlinks are rewritten to target the intended final destination of the copied unit files.
+
+        """
+        for unit_name in self.unit_names(new_wants_dir):
+            wants_symlink_path = os.path.join(new_wants_dir, unit_name)
+            package_file_path = os.path.realpath(wants_symlink_path)
+            systemd_file_path = os.path.join(self.__base_systemd, unit_name)
+            tmp_systemd_file_path = systemd_file_path + self.new_unit_suffix
+
+            # Copy the unit file to the systemd directory with a suffix added to the filename.
+            # This file will be moved to systemd_file_path when the new package set is swapped in.
+            shutil.copyfile(package_file_path, tmp_systemd_file_path)
+            shutil.copymode(package_file_path, tmp_systemd_file_path)
+
+            # Rewrite the symlink to point to the copied unit file's destination.
+            # This symlink won't point to the correct file until the copied unit file is moved to its target location
+            # during activate_new_unit_files().
+            os.remove(wants_symlink_path)
+            os.symlink(systemd_file_path, wants_symlink_path)
+
+    def remove_unit_files(self):
+        if not os.path.exists(self.__unit_directory):
+            return
+
+        for unit_name in self.unit_names(self.__unit_directory):
+            try:
+                os.remove(os.path.join(self.__base_systemd, unit_name))
+            except FileNotFoundError:
+                pass
+
+    def activate_new_unit_files(self):
+        """Move new unit files to their final locations."""
+        if not os.path.exists(self.__unit_directory):
+            return
+
+        self.remove_unit_files()
+
+        for unit_name in self.unit_names(self.__unit_directory):
+            systemd_file_path = os.path.join(self.__base_systemd, unit_name)
+            os.rename(systemd_file_path + self.new_unit_suffix, systemd_file_path)
```

Comment:
nit: maybe use `os.path.join` instead of `+` ?

Reply:
This is adding an extension to a filename, not joining parts of a path.